### PR TITLE
Add get node names method to Node class

### PR DESCRIPTION
**Pull request by ivanpauno**
Depends on https://github.com/osrf/ros2_java/pull/30.

Maybe we should target all PRs to galactic here, what do you think @jacobperron ?
We can later create PRs in all the branches that apply when contributing upstream.

### CODE REVIEW

**Comment by ivanpauno** [comment]
Converting to draft until https://github.com/osrf/ros2_java/pull/30 is merged.

**Comment by jacobperron** [comment]
> Maybe we should target all PRs to galactic here, what do you think @jacobperron ?
We can later create PRs in all the branches that apply when contributing upstream.

Sounds good to me :+1:

File: rcljava/include/org_ros2_rcljava_graph_NodeNameInfo.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+/*
+ * Class:     org_ros2_rcljava_graph_NodeName
+ * Method:    nativeFromRCL
+ * Signature: (J)V
+ */
+JNIEXPORT void
+JNICALL Java_org_ros2_rcljava_graph_NodeName_nativeFromRCL(
```
Comment (jacobperron) [inline rcljava/include/org_ros2_rcljava_graph_NodeNameInfo.h]:
```suggestion
JNICALL Java_org_ros2_rcljava_graph_NodeNameInfo_nativeFromRCL(
```

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#include <jni.h>
+#include <limits>
+
+#include "rcl/graph.h"
+#include "rcljava_common/exceptions.hpp"
+
+using rcljava_common::exceptions::rcljava_throw_exception;
+
+JNIEXPORT void JNICALL
+JNICALL Java_org_ros2_rcljava_graph_NodeName_nativeFromRCL(
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp]:
```suggestion
JNICALL Java_org_ros2_rcljava_graph_NodeNameInfo_nativeFromRCL(
```

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  JNIEnv * env, jobject self, jlong name_handle, jlong namespace_handle, jlong enclave_handle)
+{
+  const char * name = reinterpret_cast<const char *>(name_handle);
+  const char * n_namespace = reinterpret_cast<const char *>(namespace_handle);
+  const char * enclave = reinterpret_cast<const char *>(enclave_handle);
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp]: Should we guard against null handles? I don't know how this native method is intended to be used within the context of this PR.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp [hunk 4]
```diff
@@ -254,6 +254,82 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeCreateTimerHandle(
...
+    &node_namespaces,
+    &enclaves);
+  RCLJAVA_COMMON_THROW_FROM_RCL(env, ret, "rcl_get_node_names_with_enclaves failed");
+
+  if (node_names.size != node_namespaces.size || node_names.size != enclaves.size) {
+    rcljava_throw_exception(
+      env,
+      "java/lang/IllegalStateException",
+      "names, namespaces and enclaves array leghts don't match");
+    goto cleanup;
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Now that we have the dependency on `rcpputils`, I guess we can used a scoped exit for cleanup.

File: rcljava/src/main/java/org/ros2/rcljava/graph/NodeNameInfo.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    this.name = name;
+    this.namespace = namespace;
+    this.enclave = enclave;
+  }
+
+  /// Default constructor, used from jni.
+  private NodeNameInfo() {}
+
+  /// Used to create a NodeNameInfo object from jni code.
+  private native final void nativeFromRCL(
```
Comment (jacobperron) [inline rcljava/src/main/java/org/ros2/rcljava/graph/NodeNameInfo.java]: nit: it doesn't look like this method is used. Can it be removed?

Comment (ivanpauno) [reply rcljava/src/main/java/org/ros2/rcljava/graph/NodeNameInfo.java]: [Done](https://github.com/osrf/ros2_java/pull/31/commits/8734ea82b61fac64d1656169862cfc0577d6d879)!
I made the [same mistake](https://github.com/osrf/ros2_java/pull/32#discussion_r485678343) twice :man_facepalming: .

File: rcljava/include/org_ros2_rcljava_graph_NodeNameInfo.h (hunk 1 above)
Comment (ivanpauno) [reply rcljava/include/org_ros2_rcljava_graph_NodeNameInfo.h]: This code is gone, see https://github.com/osrf/ros2_java/pull/31/commits/51de623a8da281d95951795ac0af9be48fc23bf6

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp (hunk 2 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp]: removed https://github.com/osrf/ros2_java/pull/31/commits/8734ea82b61fac64d1656169862cfc0577d6d879

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp (hunk 3 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_graph_NodeNameInfo.cpp]: Code was removed https://github.com/osrf/ros2_java/pull/31/commits/8734ea82b61fac64d1656169862cfc0577d6d879

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp (hunk 4 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Done in https://github.com/osrf/ros2_java/pull/31/commits/5af448a6c06d3282acaaf24b1fe8b2a38fe924a8.

**Review by jacobperron — APPROVED** [review]

**Pull request merged**